### ISSUE TITLE
What to be expected from DeepPartial<Array<T>>?

### ISSUE THREAD

**Issue by lucifer1004** [krzkaczor/ts-essentials#112]
```typescript
type Student = Array<{name: string, age: number}>
type PartialStudent = Partial<Student> 
// this yields ({
//    name: string;
//    age: number;
//} | undefined)[]
type DeepPartialStudent = DeepPartial<Student>
// this yields ({
//    name?: string;
//    age?: number;
//} | undefined)[]
```

Is this to be expected? Applying `Partial` or `DeepPartial` on an `Array` allows array elements to be `undefined`? @krzkaczor @quezak

**Comment by lucifer1004** [comment]
The advantage of current implementation (not handling `Array` as a special case) is that `tuple` can be handled neatly. 
If `Array` is handled like `Map` or `Set`, Deep-* on a tuple will not work as expected.

**Comment by quezak** [comment]
The intuitive way for me would be that `DeepPartial<Array<T>>` gives you just `Array<DeepPartial<T>>`.

Let's see what others have to say about tuples. Personally for me, if you need a deep partial on a tuple with different element types, it's probably a sign of bad code design, and the object should be an actual object with named keys instead.

/summon @macbem

**Comment by krzkaczor** [comment]
I also find that DeepPartial<array> is gonna produce array of undefines at least weird. 

I would be happy to change this behavior and release as breaking change OR maybe introducing a new type? Also, should be there a consistency with other `Deep-*` types?

**Comment by lucifer1004** [comment]
If we are to change this behavior, I think all Deep-* types should be changed.

**Comment by krzkaczor** [comment]
Released as `ts-essentials@5.0.0`
